### PR TITLE
27 enable interpolatedcurve to take any interpolation type

### LINKED ISSUE #27: Enable InterpolatedCurve to take any interpolation type.
**Context**
The interpolation curve can currently only do the LinearSpline interpolation method. This limits the flexibility of the class in terms of difficulty to test other interpolation methods.

**Details**
- [ ] Introduce an abstraction to generalize the interpolation method.
- [ ] Modify InterpolatedCurve to accept any implementation of this interface.
- [ ] Update class to remove hardcoded "LinearSpline"

### PR DESCRIPTION
**Pull request by rwx-oberholzer-marcelle**
✅ **Pull Request Checklist**

### 🔍 Self-Review

* [x] Destination branch is correct.
* [x] Code has been self-reviewed and is scoped to the PR’s purpose.
* [x] No debugging code, print statements, or unrelated changes.
* [x] Code is clean, well-commented, and follows project standards.
* [x] Documentation is updated, if applicable.

### 🧪 Testing & Build

* [ ] All relevant tests (unit/integration) pass locally.
* [ ] Edge cases and error handling are considered.
* [ ] No new warnings or errors in the build or test output.

### 📄 Clarity

* [ ] PR title and description are clear and descriptive.
* [ ] Purpose and context of the change are well explained.
* [ ] Related issues are referenced (e.g., `Fixes #123`).

### 👥 Collaboration

* [x] Reviewers assigned and feedback requested.
* [x] All review comments have been addressed or responded to.

### CODE REVIEW

File: QuantSA/QuantSA.Core/CurvesAndSurfaces/InterpolationAdapters/InterpolationAdapters.cs [hunk 1]
```diff
@@ -0,0 +1,19 @@
+﻿using MathNet.Numerics.Interpolation;
+
+namespace QuantSA.Core.CurvesAndSurfaces
+{
+    public class LinearSplineAdapter : IInterpolator1D
```
Comment (rwx-jacquier-robert) [inline QuantSA/QuantSA.Core/CurvesAndSurfaces/InterpolationAdapters/InterpolationAdapters.cs]: You can remove the "Adapter" part from the name.

File: QuantSA/QuantSA.Core/CurvesAndSurfaces/InterpolationAdapters/InterpolationAdapters.cs [hunk 2]
```diff
@@ -0,0 +1,19 @@
...
+    {
+        private readonly LinearSpline _spline;
+
+        public LinearSplineAdapter(double[] xVals, double[] yVals)
+        {
+            _spline = LinearSpline.InterpolateSorted(xVals, yVals);
```
Comment (rwx-jacquier-robert) [inline QuantSA/QuantSA.Core/CurvesAndSurfaces/InterpolationAdapters/InterpolationAdapters.cs]: If you make the change listed above you'll likely have to call the MathNet LinearSpline by its full name.

**Review by rwx-jacquier-robert — CHANGES_REQUESTED** [review]
I really like the interface you introduced.

Don't forget to add/adjust tests for the class.

File: QuantSA/QuantSA.Core.Tests/CurvesAndSurfaces/InterpolationAdapters/InterpolationAdaptersTests.cs
Comment (rwx-jacquier-robert) [inline QuantSA/QuantSA.Core.Tests/CurvesAndSurfaces/InterpolationAdapters/InterpolationAdaptersTests.cs:1]: You probably want to add tests that lie on the left and right of the boundaries so the behaviour is understood.

**Review by rwx-jacquier-robert — CHANGES_REQUESTED** [review]

File: QuantSA/QuantSA.Core.Tests/CurvesAndSurfaces/InterpolationAdapters/InterpolationAdaptersTests.cs [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+        {
+            var curve = new InterpolatedCurve(_xVals, _yVals);
+
+            double valueAt1_5 = curve.Interp(1.5);
+
+            Assert.AreEqual(15, valueAt1_5, 1e-10, "Interpolation failed");
+        }
+
+        [TestMethod]
+        public void InterpolatedCurve_ExtrapolatesLeft_ReturnsASExpected()
```
Comment (rwx-jacquier-robert) [inline QuantSA/QuantSA.Core.Tests/CurvesAndSurfaces/InterpolationAdapters/InterpolationAdaptersTests.cs]: Spelling AS

File: QuantSA/QuantSA.Core.Tests/CurvesAndSurfaces/InterpolationAdapters/InterpolationAdaptersTests.cs [hunk 4]
```diff
@@ -0,0 +1,49 @@
...
+        private double[] _yVals;
+
+        [TestInitialize]
+        public void Setup()
+        {
+            _xVals = new double[] { 1, 2, 3 };
+            _yVals = new double[] { 10, 20, 30 };
+        }
+        [TestMethod]
+        public void InterpolatedCurve_ReturnsValues_AsExpected()
```
Comment (rwx-jacquier-robert) [inline QuantSA/QuantSA.Core.Tests/CurvesAndSurfaces/InterpolationAdapters/InterpolationAdaptersTests.cs]: Test naming "Interp_ReturnsExpectedValues"

**Review by rwx-jacquier-robert — CHANGES_REQUESTED** [review]
Revisit all test names to match naming conventions.

File: QuantSA/QuantSA.Core.Tests/CurvesAndSurfaces/InterpolationAdapters/InterpolationAdaptersTests.cs [hunk 5]
```diff
@@ -0,0 +1,49 @@
...
+        {
+            var curve = new InterpolatedCurve(_xVals, _yVals);
+
+            double valueAt1_5 = curve.Interp(1.5);
+
+            Assert.AreEqual(15, valueAt1_5, 1e-10, "Interpolation failed");
+        }
+
+        [TestMethod]
+        public void InterpolatedCurve_ExtrapolatesLeft_ReturnsAsExpected()
```
Comment (rwx-jacquier-robert) [inline QuantSA/QuantSA.Core.Tests/CurvesAndSurfaces/InterpolationAdapters/InterpolationAdaptersTests.cs]: ReturnsExpectedValue is better.

File: QuantSA/QuantSA.Core.Tests/CurvesAndSurfaces/InterpolationAdapters/InterpolationAdaptersTests.cs [hunk 6]
```diff
@@ -0,0 +1,49 @@
...
+        private double[] _yVals;
+
+        [TestInitialize]
+        public void Setup()
+        {
+            _xVals = new double[] { 1, 2, 3 };
+            _yVals = new double[] { 10, 20, 30 };
+        }
+
+        [TestMethod] public void Interp_ReturnsExpectedValues()
```
Comment (rwx-jacquier-robert) [inline QuantSA/QuantSA.Core.Tests/CurvesAndSurfaces/InterpolationAdapters/InterpolationAdaptersTests.cs]: ReturnsExpectedValue singular.

File: QuantSA/QuantSA.Core.Tests/CurvesAndSurfaces/InterpolationAdapters/InterpolationAdaptersTests.cs [hunk 7]
```diff
@@ -0,0 +1,49 @@
...
+        {
+            var curve = new InterpolatedCurve(_xVals, _yVals);
+
+            double valueAt1_5 = curve.Interp(1.5);
```
Comment (rwx-jacquier-robert) [inline QuantSA/QuantSA.Core.Tests/CurvesAndSurfaces/InterpolationAdapters/InterpolationAdaptersTests.cs]: just name the variable "result".

**Review by rwx-jacquier-robert — CHANGES_REQUESTED** [review]
Apply the comments to all test names/variables.

**Comment by rwx-jacquier-robert** [comment]
@MarcelleOberholzer also convert this from a draft to an actual PR.
